ENH: sparse.linalg: condition number for sparse matrices

broesler opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes https://github.com/scipy/scipy/issues/18969
Builds on PR #21620, which is currently open.

#### What does this implement/fix?
This PR is a continuation of the work done in PR #21620, which introduced the
`scipy.sparse.linalg.cond1est` function, and the
`scipy.sparse.linalg.rinvnormest` function.

Changes herein:
- Retain `cond1est`
- Remove `rinvnormest` from `scipy.sparse.linalg`
- Refactor `rinvnormest` to `normest_inv` in `scipy.sparse.linalg.SuperLU`


#### Additional information
The `cond1est` function estimates the condition number of a sparse matrix in the
1-norm, which is defined as the product of the 1-norm of the matrix, and the
1-norm of its inverse. This PR slightly changes the internal functionality of
`cond1est` by computing the exact 1-norm of the matrix with
`sparse.linalg.norm`, as opposed to the estimate `sparse.linalg.onenormest`.
This PR also adds input checking to `cond1est`, and updates the docstring.

This PR removes the `scipy.sparse.linalg.rinvnormest` function from
`scipy.sparse.linalg`, but retains the functionality in the
`scipy.sparse.linalg.SuperLU` class. There is no analogous function in either
numpy or scipy for dense matrices, so it is unnecessary to expose it at the top
level of `scipy.sparse.linalg`.

This PR renames the `SuperLU.rinvnormest` method to `normest_inv`, which better
reflects the mathematical syntax ``norm(inv(A))``. The method now returns the
norm itself rather than its reciprocal. The API has also been updated to accept
either the integer `1` or float `np.inf` as the norm order, as opposed to
a string argument.

Finally, this PR adds extensive tests for the `normest_inv` and `cond1est` for
sparse matrices of various sizes, densities, and data types. The tests also
include checks for the correct return types and values, as well as error
handling for invalid inputs.

There is an additional script in
`scipy/sparse/linalg/_dev_scripts/check_normcond_behavior.py` that demonstrates
the existing behavior of `np.linalg.norm`, `np.linalg.cond`,
`scipy.linalg.norm`, and `scipy.sparse.linalg.onenormest`.

maxaehle committed:
ENH: add scipy.sparse.linalg.SuperLU.rinvnormest

Implemented in the C layer, calls the SuperLU gscon function.

maxaehle committed:
ENH: add scipy.sparse.linalg.rinvnormest

Calls scipy.sparse.linalg.splu and then calls the rinvnorm
function of the resulting LU decomposition.

maxaehle committed:
ENH: add scipy.sparse.linalg.cond1est

using scipy.sparse.linalg.rinvnormest and ...onenormest

maxaehle committed:
TST: Add rinvnormest related tests

maxaehle committed:
BUG: scipy.sparse.linalg.cond1est returned reciprocal

instead of the condition number itself

maxaehle committed:
TST: Add test for scipy.sparse.linalg.cond1est

maxaehle committed:
TST: Test if ValueEror is raised in rinvnormest

maxaehle committed:
BUG/TST: test sparse rinvnormest with actual float/complex matrices

broesler committed:
ENH: Rename rinvnormest -> invnormest. Return norm itself, not reciprocal.

Returning 1 / norm(inv(A)) is cumbersome. Now we just return
norm(inv(A)).

In addition, rename the module to _cond1est to reflect the actual
intended enhancement.

Also, add parameter names to gscon functions in SuperLU header.

broesler committed:
WIP: Update tests and add notes for future TODOs.

broesler committed:
BLD: Fix meson.build files for cond1est rename.

broesler committed:
ENH: Change API of invnormest to accept 1 or np.inf.

The API now mirrors that of scipy.sparse.linalg.norm(A, ord=ord), with
the caveat that invnormest only has 2 options for `ord` {1, np.inf},
since the underlying SuperLU `gscon` function only supports those
options.

broesler committed:
ENH: Compute exact 1-norm of A in cond1est.

Needs performance testing, but the condition number estimate will be
more accurate if we are not multiplying two estimates together. The
sparse 1-norm computation of the original matrix should be efficient.

Note that the __init__.py imports needed to be reordered so that
scipy.linalg.norm is available in cond1est.

broesler committed:
ENH: Add error checking to cond1est.

broesler committed:
MNT: Remove TODO comment.

broesler committed:
DOC: update docstrings with See Also, Notes, References.

Also update doctests to match normest_inv name, and add versionadded to
docstrings.

broesler committed:
ENH: Inline temporary variable for LU decomposition.

broesler committed:
ENH: Refactor invnormest -> normest_inv.

This rename of the SuperLU.normest_inv method is more consistent with
the way that the mathematical function would be written
``norm(inv(A))``, where "est" is like a subscript for "estimate".

broesler committed:
ENH: Remove normest_inv as a public function in sparse.linalg.

It is unnecessary to expose it, as there is no counterpart in either
dense library numpy.linalg or scipy.linalg. Currently, its only use is
in ``sparse.linalg.cond1est``. The method is still available in the
SuperLU object.

broesler committed:
DOC: Change csc_matrix to csc_array in cond1est docstring.

broesler committed:
TST: Refactor tests to separate normest_inv and cond1est classes.

Also add large comment block describing existing behavior of numpy and
scipy for similar functions.

broesler committed:
ENH: Return the appropriate np.float dtype from normest_inv.

broesler committed:
MAINT: refactor cond -> norm for clarity of intent.

Also reorganize the variable declarations to be as close as possible to
their actual use. Maintain volatile where appropriate for thread-safe
use.

broesler committed:
DEV: Create script to test existing norm/cond behavior.

This script is not part of the scipy package itself, but demonstrates
the behavior of existing norm and cond functions so that the new
normest_inv and cond1est can return appropriate values and raise
appropriate warnings and errors.

broesler committed:
TST: Overhaul tests for normest_inv and cond1est.

broesler committed:
TST: Fix up dtype and norm_ord parametrization.

broesler committed:
BUG: Add check for empty matrix in normest_inv.

broesler committed:
MNT: Write return condition as a boolean variable.

broesler committed:
MNT: Refactor empty matrix check to a static function.

broesler committed:
ENH: Raise TypeError on non-sparse input to cond1est.

Also add associated test. This change makes cond1est behave consistently
with sparse.linalg.norm.

broesler committed:
TST: Test with complex-valued matrices.

See also the NOTE in the comments regarding the np.linalg.cond bug. It
returns complex values when the matrix dtype is complex, whereas the
condition number should always be a real.

broesler committed:
TST: Refactor SEED to rng instance.

This commit decouples any test fixtures called at the same time with the
same seed. It revealed a bug in generate_matrix where a given "exactly
singular" matrix may raise a different error from splu, so this commit
also updates the construction of the random matrices to ensure we get an
exactly singular matrix when requested.

broesler committed:
TST: Refactor return_dtype to a fixture. Test all return types.

broesler committed:
TST: Change sampler to normal distribution so we get negative values.

The default uniform [0, 1) only gives positive matrix values, whereas
N(0, 1) gives both positive and negative values. This change
tests to ensure negative values work properly.

broesler committed:
TST: Refactor norm variable names.

broesler committed:
TST: Parametrize tests for normest_inv and cond1est by N and density.

Run these tests over 10 trails per data type (and norm_ord). Since the
matrices are randomly generated, try a number of samples to ensure we
didn't get lucky one a single sample.

broesler committed:
TST: Fix bug in testing with rng.normal sampler.

broesler committed:
MNT: Clean up comments.

broesler committed:
DOC: Update docstrings.

Fixes include:
- More informative citation.
- Fix doctest output in SuperLU.normest_inv.
- Update sparse/linalg/__init__.py docstring to include cond1est.
- Style tweaks to the documentation.

ilayn commented:
This is really impressive work but before we go further, note the discussion on #21620 until the end. 

We don't really need the norm of the inverse but we need a condition number which is theoretically defined as $\kappa$ but typically it is not derived using separate two norm estimate products. That's why we were discussing `lacon2` calls in the end and that's still need to be done.

broesler commented:
Ok I was going to ask if that is still under consideration.

I looked into the SuperLU function calls, and I think the best way to achieve that goal is actually to use the [langs](https://portal.nersc.gov/project/sparse/superlu/superlu_code_html/dlangs_8c.html) function to compute the exact 1-norm of the original `A` (in C), then calling `gscon` on L and U to estimate the condition number. The referenced Hager and Higham papers both discuss computing the 1-norm of A exactly, and estimating the 1-norm of A inverse. If we use `lacon2` to estimate the 1-norm of A from L and U, we are essentially just multiplying L and U back together to get the original matrix, which is inefficient.

The caveat here is that the 1-norm computation in `langs` would have to be done during the *construction* of the SuperLU object, while we still have the original matrix (as opposed to just the L and U factors). That should be a cheap operation relative to the factorization, but I can add testing to make sure that is the case.

broesler committed:
ENH: Compute cond1est as SuperLU member function. Remove normest_inv.

This commit addressed reviewer comments on PR #23292 regarding internal
computation of `cond1est` vs. computing two norms in Python.

This commit makes an internal change to the SuperLU object to compute
and store the 1-norm of the matrix A when splu(A) is called. That value
is then used by the new member function `SuperLU.cond1est`, which calls
`gscon` to compute the condition number.

This commit removes the existing `SuperLU.normest_inv` member function,
which is no longer necessary. It also removes the associated test class
TestNormEstInv.

broesler committed:
DOC: Fix input type of `A` to `cond1est` to match actual behavior.

broesler committed:
TST: Update test types to generic single and double precision types.

Use, e.g., np.single instead of np.float32, and the respective complex
types. Also increase the relative tolerance for result comparison
depending on whether the dtype is single or double precision.

broesler commented:
Hi, just checking in on the status of this PR. Let me know if you need me to make any more changes before merging.

ilayn commented on scipy/sparse/linalg/_dev_scripts/check_normcond_behavior.py:
```diff
@@ -0,0 +1,295 @@
+#!/usr/bin/env python3
```
This file needs to be converted to actual tests such that they always run in case they are not exact copies of the tests below. We don't keep additional development versions of the designs.

ilayn commented on scipy/sparse/linalg/_dsolve/_superluobject.c:
```diff
@@ -105,10 +106,159 @@ static PyObject *SuperLU_solve(SuperLUObject * self, PyObject * args,
+        if (ord == -1.0 && PyErr_Occurred()) {
+            return NULL;  /* conversion error */
+        }
+        if (ord == NPY_INFINITY) {
```
I think norm1est is OK with getting only the 1-norm so you don't need to perform this norm detection dance. Infinity norm can be obtained by the transpose of the sparse array anyways.

ilayn commented on scipy/sparse/linalg/_dsolve/_superluobject.c:
```diff
@@ -105,10 +106,159 @@ static PyObject *SuperLU_solve(SuperLUObject * self, PyObject * args,
+        return NULL;
+    }
+
+    const bool return_float = (self->type == NPY_FLOAT || self->type == NPY_CFLOAT);
```
Use dtype NPY_FLOAT32 and NPY_COMPLEX64 for explicit bit width specs

ilayn commented on scipy/sparse/linalg/_dsolve/_superluobject.c:
```diff
@@ -105,10 +106,159 @@ static PyObject *SuperLU_solve(SuperLUObject * self, PyObject * args,
+    volatile int info;
+
+    switch (self->type) {
+        case NPY_FLOAT:
```
Explicit NumPy dtypes here too.

broesler replied:
Explicit dtypes used in 47caff508.

ilayn commented on scipy/sparse/linalg/_dsolve/_superluobject.c:
```diff
@@ -105,10 +106,159 @@ static PyObject *SuperLU_solve(SuperLUObject * self, PyObject * args,
+
+    SLU_END_THREADS;
+
+    if (info) {
```
needs an explicit check for `info < 0` in case something goes weird and somehow info comes back positive overflow or whatever the case might be because ... C.

ilayn commented on scipy/sparse/linalg/_dsolve/_superluobject.c:
```diff
@@ -721,6 +871,26 @@ PyObject *newSuperLUObject(SuperMatrix * A, PyObject * option_dict,
     self->py_csc_construct_func = NULL;
     self->type = intype;
 
+    /* Call the appropriate function to compute the 1-norm of the matrix */
```
This has to be a lazy evaluation otherwise every instance will have to have the norm computation even if they won't need it. Hence you have to call a class method inside the `norm1est` to get the norm.

broesler replied:
For reference, I benchmarked the code on my local laptop. Here is a comparison of the addition of the norm computation into the SuperLU object creation:

```
% asv compare main feature_cond1est

All benchmarks:

| Change   | Before [601dc92c] <main>   | After [7572f59e] <feature_cond1est>   |   Ratio | Benchmark (Parameter)                                          |
|----------|----------------------------|--------------------------------------------|---------|----------------------------------------------------------------|
|          | 285M                       | 286M                                       |    1    | sparse_linalg_cond1est.Norm1EstBench.peakmem_splu(5000, 0.001) |
|          | 501M                       | 499M                                       |    1    | sparse_linalg_cond1est.Norm1EstBench.peakmem_splu(5000, 0.1)   |
|          | 62.5M                      | 62.8M                                      |    1    | sparse_linalg_cond1est.Norm1EstBench.peakmem_splu(5000, 1e-05) |
|          | 2.69±0.02s                 | 2.70±0.02s                                 |    1    | sparse_linalg_cond1est.Norm1EstBench.time_splu(5000, 0.001)    |
|          | 7.95±0.01s                 | 8.00±0.04s                                 |    1.01 | sparse_linalg_cond1est.Norm1EstBench.time_splu(5000, 0.1)      |
|          | 313±10μs                   | 318±2μs                                    |    1.02 | sparse_linalg_cond1est.Norm1EstBench.time_splu(5000, 1e-05)    |
```

The norm computation adds 1-2% additional time in constructing the SuperLU object for a 5000 x 5000 matrix. 

Here is a comparison of the `cond1est` computation time using the C computation of the norm vs. a `sparse.linalg.norm` call:

```
% asv compare feature_cond1est feature_cond1est_pynorm

All benchmarks:

| Change   | Before [7572f59e] <feature_cond1est>   | After [ef18179e] <feature_cond1est_pynorm>   |   Ratio | Benchmark (Parameter)                                              |
|----------|----------------------------------------|----------------------------------------------|---------|--------------------------------------------------------------------|
|          | 295M                                   | 286M                                         |    0.97 | sparse_linalg_cond1est.Norm1EstBench.peakmem_cond1est(5000, 0.001) |
|          | 502M                                   | 520M                                         |    1.04 | sparse_linalg_cond1est.Norm1EstBench.peakmem_cond1est(5000, 0.1)   |
|          | 64.9M                                  | 63.6M                                        |    0.98 | sparse_linalg_cond1est.Norm1EstBench.peakmem_cond1est(5000, 1e-05) |
|          | 2.75±0.04s                             | 2.76±0.02s                                   |    1    | sparse_linalg_cond1est.Norm1EstBench.time_cond1est(5000, 0.001)    |
|          | 8.38±0.1s                              | 8.19±0.01s                                   |    0.98 | sparse_linalg_cond1est.Norm1EstBench.time_cond1est(5000, 0.1)      |
| +        | 451±10μs                               | 574±30μs                                     |    1.27 | sparse_linalg_cond1est.Norm1EstBench.time_cond1est(5000, 1e-05)    |
```

There is a bit of noise in the least dense matrix since there are only ~250 entries, but the denser matrices show that the methods are essentially equivalent in both time and memory usage. We expect this result since the norm computation is almost negligible vs. the LU factorization itself.

broesler replied:
Comparison of the branches is here:

[Branch comparison](https://github.com/broesler/scipy/compare/feature_cond1est...feature_cond1est_pynorm)

I will merge the `feature_cond1est_pynorm` version with the `feature_cond1est` branch in this PR if we would like to use that one, but I wanted to make sure the reviewers could see both without opening an additional PR.

broesler replied:
I realize 5,000 x 5,000 isn't particularly large, but 10,000 x 10,000 timed out the `asv` run on my local machine (60s max). In running a simple test with a smaller random matrix, I noticed that `splu` is quite slow relative to the dense `lu`:

```python
>>> from scipy import sparse, linalg as la
>>> from scipy.sparse import linalg as spla
>>> N = 5000
>>> d = 0.001
>>> A = sparse.random_array((N, N), density=d, format='csc')
>>> A.setdiag(N)  # set diagonal large to avoid pivoting
>>> A
===
<Compressed Sparse Column sparse array of dtype 'float64'
        with 29997 stored elements and shape (5000, 5000)>
>>> %timeit la.lu(A.toarray())
609 ms ± 10.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit spla.splu(A)
2.74 s ± 3.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

The sparse version is ~4.4x slower than the dense version.

This difference reverses using a structured matrix:

```python
>>> from scipy.sparse.linalg import LaplacianNd
>>> sqrtN = 71  # 71 * 71 = 5041 ~ 5000
>>> A = LaplacianNd((sqrtN, sqrtN)).tosparse().tocsc().astype(float)
>>> A
===
<Compressed Sparse Column sparse array of dtype 'float64'
        with 24921 stored elements and shape (5041, 5041)>
>>> %timeit la.lu(A.toarray())
3.71 s ± 1.36 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit spla.splu(A)
34.3 ms ± 497 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Each of these matrices has essentially the same density, so I am not quite sure why there are such major runtime discrepancies.

ilayn commented on scipy/sparse/linalg/_dsolve/_superluobject.c:
```diff
@@ -721,6 +871,26 @@ PyObject *newSuperLUObject(SuperMatrix * A, PyObject * option_dict,
+    /* Call the appropriate function to compute the 1-norm of the matrix */
+    char norm_c = '1';  /* default to 1-norm */
+    switch (self->type) {
+        case NPY_FLOAT:
```
It is better to spell out the bit width for more explicit dtypes in case some architecture changes these, that is NPY_FLOAT64, NPY_COMPLEX128 etc. even though they are typically aliases.

ilayn commented on scipy/sparse/linalg/_dsolve/_superluobject.h:
```diff
@@ -25,6 +26,20 @@
+ * It is either a float or a double, depending on the type of the matrix.
+ * This union allows us to use the same variable for both types.
+ */
+typedef union _SuperLUAnorm_t {
```
This is better to have a struct otherwise you are bit fiddling for 4 bytes. There is no need for such frugality here and for type punning.

broesler replied:
struct used in 4497567c0.

ilayn reviewed (changes requested):
This looks quite good. Thanks @broesler . I've added some comments.

broesler committed:
MNT: Remove _dev_scripts/check_normcond_behavior.py.

This commit removes a development script from the repo in response to
reviewer comment. It is auxiliary to the existing unit tests.

broesler committed:
MNT: Remove "ord" kwarg from SuperLU.cond1est method.

This commit is in response to a reviewer comment. This kwarg is
unnecessary, since the name "cond1est" implies use of the 1-norm, and
the `scipy.sparse.linalg.cond1est` function that calls
this method does not support the kwarg.

broesler committed:
MNT: Use explicit bit widths in SuperLU_cond1est.

This commit is in response to a reviewer comment. It replaces:
* NPY_FLOAT -> NPY_FLOAT32
* NPY_DOUBLE -> NPY_FLOAT64
* NPY_CFLOAT -> NPY_COMPLEX64
* NPY_CDOUBLE -> NPY_COMPLEX128

broesler committed:
MNT: Use struct instead of union for SuperLUAnorm_t.

This commit is in response to a reviewer comment for the change.

broesler committed:
MNT: Add explicit check for info < 0.

This commit is in response to a reviewer comment.
The SuperLU documentation specifies the `info` parameter values:

  INFO    (output) int*
          = 0:  successful exit
          < 0:  if INFO = -i, the i-th argument had an illegal value

so we check explicitly for `info < 0` and assume success otherwise.

broesler committed:
ENH/BENCH: Remove 1-norm computation from SuperLU object creation.

This commit effectively reverts to b3c80ed, but includes intermediate
changes based on reviewer comments, so it is not actually a "revert"
commit. Also includes minor documentation updates.

We now compute the 1-norm of the input matrix A using
``sparse.linalg.norm`` instead of internally in the SuperLU object.
Thus, the norm will only be computed when `cond1est` is called, not for
every SuperLU object that is created.

This commit also includse a benchmark test for comparison with the
internal C norm computation version.

broesler committed:
BENCH: Add benchmark test for cond1est/splu.

broesler committed:
BENCH: Use LaplacianNd in cond1est benchmark.

Using a random_array causes splu to run quite slowly for large matrices
(N > 5000), so asv times out at 60s. A LaplacianNd array is much faster,
so we can run the benchmark on larger matrices.

broesler commented:
Ah, the `strict=True` kwarg in `np.testing.assert_allclose` was introduced in numpy v2.0.0, so all of the failures are due to the test being run with an earlier version of numpy. I see that the `scipy/pyproject.toml` requires '"numpy>=2.0.0"', but do these CI runs just override that requirement?

broesler committed:
TST: Remove `strict=True` in test_cond1est.py.

This kwarg is only compatible with numpy>=2.0.0, so this commit rewrites
the tests to be compatible with older versions of numpy.

broesler commented:
Agreed, the MacOS fix seems unrelated to this PR.

The Windows version seems to be failing in `np.linalg.cond`, which is the first line of the failing test. See lines 1078-1079 of the "Test" output in the Windows CI job:

```
 File "C:\hostedtoolcache\windows\Python\3.12.10\arm64\Lib\site-packages\numpy\linalg\_linalg.py", line 2035 in cond
  File "C:\a\scipy\scipy\build-install\Lib\site-packages\scipy\sparse\linalg\tests\test_cond1est.py", line 257 in test_random_nonsingular_matrix
```

so it is not even getting to the SuperLU calls. This issue might also be fixed by rebasing on main?

broesler committed:
REV: TST: Update test types to generic single and double precision types.

This reverts commit b836a8e0f1bb19b2bd962d5f6bd0a3ca214271c0. This
commit is an attempt to figure out the bug in WIN-ARM64 parallel
processing testing.
See [Numpy Issue #30089](https://github.com/numpy/numpy/issues/30089).
